pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java

```diff
@@ -54,14 +55,27 @@ public ControllerPeriodicTask(String taskName, long runFrequencyInSeconds, long
     _controllerMetrics = controllerMetrics;
   }
 
+  // Determine if this task can run on the specified table. Task can run on all tables for which the controller is lead
+  // if "tablename" property is not set. However, if "tablename" property is set (by calling the /periodictask/run
+  // controller API), then the task will only run on the specified by the "tablename" property key.
+  private boolean shouldRunTaskForTable(String tableNameWithType) {
+    if (_leadControllerManager.isLeaderForTable(tableNameWithType)) {
+      String propTableNameWithType = (String) _activePeriodicTaskProperties.get(PeriodicTask.PROPERTY_KEY_TABLE_NAME);
+      return propTableNameWithType == null || propTableNameWithType.equals(tableNameWithType);
+    }
+    return false;
+  }
+
   @Override
   protected final void runTask() {
     _controllerMetrics.addMeteredTableValue(_taskName, ControllerMeter.CONTROLLER_PERIODIC_TASK_RUN, 1L);
     try {
       // Process the tables that are managed by this controller
+      // TODO: creating tablesToProcess list below is redundant since processTables will unroll the list anyway. This
+      // needs to be cleaned up sometime.
       List<String> tablesToProcess = new ArrayList<>();
       for (String tableNameWithType : _pinotHelixResourceManager.getAllTables()) {
-        if (_leadControllerManager.isLeaderForTable(tableNameWithType)) {
+        if (shouldRunTaskForTable(tableNameWithType)) {
```

Comment:
This will be very inefficient when table name is already provided. We should only check for the provided table name instead.
If we have a separate method `runTask(Properties periodicTaskProperties)` it will be much cleaner

Reply:
Pending...